awx/main/scheduler/task_manager_instances.py

```diff
@@ -0,0 +1,213 @@
+        if not isinstance(controller_node, PrioritizableNode) and isinstance(controller_node, str):
+            controller_node = self.control_nodes[controller_node]
+        controller_node.remaining_capacity -= settings.AWX_CONTROL_NODE_TASK_IMPACT
+        self.control_nodes.update(controller_node)
+
+    def assign_task_to_control_node(self, task, ig_capacity_graph, controller_node=None):
+        """Account for task impact on a control node.
+
+        If no control node is specified, find the best control node available.
+        controller_node should be the string hostname
+
+        If the node does not have enough remaining capacity for the task impact, return False
+        """
+        impact = settings.AWX_CONTROL_NODE_TASK_IMPACT
+        if not controller_node:
+            instance = self.control_nodes.best_node()
+        else:
+            instance = self.control_nodes[controller_node]
+
+        if not self.has_sufficient_control_capacity(instance):
+            logger.warning(f"Not enough control capacity for task {task.log_format} with task impact {impact} found on nodes {self.control_nodes}")
+            return False
+
+        for ig in self.instances_partial[instance.hostname].instance_groups:
+            ig_capacity_graph.consume_instance_group_capacity(task, ig, instance=self.instances_partial[instance], impact=impact)
+            ig_data = ig_capacity_graph.get(ig)
+            if not ig_data['control_capacity'] - ig_data['consumed_control_capacity'] >= 0:
+                # This whole block is kind of paranoid.
+                # It covers a case that should really not happen since every time we assign a task to a node, we deduct from its instance group
+                # capacity as well. If we never see this warning, maybe we can drop this code
+                logger.warn(
+                    f"""Somehow we had capacity on a control node {instance}
+                        but not on its instance_group {ig} that has {ig_data['control_capacity']} control capacity
+                        and {ig_data['consumed_control_capacity']} consumed control capacity"""
+                )
+        logger.debug(f"chose {instance} to be control node for {task.log_format}")
+        self.consume_control_capacity(instance)
+        return instance.hostname
+
+
+class InstanceGroupCapacityGraph:
+    def __init__(self, task_manager_instances):
+        self.ig_capacity_graph = dict()
+        breakdown = False
+        for rampart_group in InstanceGroup.objects.prefetch_related('instances'):
+            InstanceGroupManager.zero_out_group(self.ig_capacity_graph, rampart_group.name, breakdown)
+            # Didn't move the init of DependencyGraph to InstanceGroupManager because of circular import
+            self.ig_capacity_graph[rampart_group.name]['dependency_graph'] = DependencyGraph()
+            for instance in rampart_group.instances.filter(enabled=True).order_by('hostname'):
```

Comment:
I'm never entirely clear on behavior with experimenting, but if you do `rampart_group.instances.filter(enabled=True).order_by('hostname')`, do you hit the cache populated by `InstanceGroup.objects.prefetch_related('instances')`?

With the complexity here... I don't think so. If you want enabled instances ordered by hostname, then you need to do that in python. The thing I know you can safely reference without performance penalties is `rampart_group.instances.all()`

Reply:
I really don't think the order matters, this is just what the code was previously doing. I'll experiment w/o the order_by